akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala

```diff
@@ -481,33 +481,33 @@ private[client] object NewHostConnectionPool {
 
             withSlot(_.debug("Received response")) // FIXME: add abbreviated info
 
-            response.entity match {
-              case _: HttpEntity.Strict => withSlot(_.onResponseReceived(response))
-              case e =>
-                val (newEntity, StreamUtils.StreamControl(entitySubscribed, entityComplete, entityKillSwitch)) =
-                  StreamUtils.transformEntityStream(response.entity, StreamUtils.CaptureMaterializationAndTerminationOp)
+            // if hasNoEntityStream == true, we don't expect onResponseEntity... events in the state machine above
+            if (hasNoEntityStream(response)) withSlot(_.onResponseReceived(response))
```

Comment:
The fix here is that we now use the same logic (`hasNoEntityStream`) here and above in the state machine to determine whether to send `onResponseEntity...` event or not, so the bug shouldn't happen any more.